presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java

```diff
@@ -189,7 +193,7 @@ private void sanityCheckFragmentedPlan(SubPlan subPlan, WarningCollector warning
         }
     }
 
-    private SubPlan analyzeGroupedExecution(Session session, SubPlan subPlan)
+    private SubPlan analyzeGroupedExecution(Session session, SubPlan subPlan, boolean parentContainsTableFinish)
```

Comment:
Add comments here (or at the global level of `PlanFragmenter`): 

```
In theory, recoverable grouped execution should be decided at query section level (i.e. 
a connected component of stages connected by remote exchanges). This is because 
there are little benefits to have mixed recoverable execution and non-recoverable execution 
within the query section (a single task failure will still like to cause the section execution failure),
but adds unnecessary complications. 

However, since "query section" is not introduced until execution time, today it will
be quite hacky to decide it at fragmenting time. 

TODO: We should introduce "plan section" and make this done prior to plan fragmenting. 
```